13th and 14th February 2020, University of Cambridge, Cambridge, UK

# Instructors

- Nick Gleadall (ng384@medschl.cam.ac.uk)
- Alba Sanchis-Juan (as2635@medschl.cam.ac.uk)
- Jonathan Stephens (Jonathan.stephens@bioresource.nihr.ac.uk)
- Luca Stefanucci (ls760@medschl.cam.ac.uk)

# Description

Analysis of whole genome data unearths a multitude of variants of different classes, which need to be filtered, annotated and validated to arrive at a causative variant for a disease. The current short length sequences, whilst being excellent at identifying single nucleotide variants and short insertions/deletions, struggle to correctly map structural variants (SVs). Long-read sequencing technologies offer improvements in the characterisation of genetic variation and regions that are difficult to assess with short-read sequences.
The aim of this course is to familiarise participants with long read sequencing technologies, their applications and the bioinformatics tools used to assemble this kind of data. Lectures will introduce this technology and provide insight into methods for the analysis of genomic data, while the hands-on sessions will allow participants to run analysis pipelines, focusing on data generated by the Oxford Nanopore Technologies (ONT) platform.

# Prerequisites

The course is suitable for complete beginners and assumes no prior programming experience.
Basic knowledge of R and UNIX would be an advantage. An introduction to the the Unix system and shell is available
<a href="https://swcarpentry.github.io/shell-novice/">here</a>

# Detailed aims

This course will provide:

- Introduction to ONT devices
- Wet lab preparation of libraries for Nanopore sequencing
- Introduction to data processing of long-reads
- Variant calling and quality control
- Structural variant calling and quality control
- Haplotype phasing

# Objectives

After this course you should be able to:

- Discuss advantages and limitations of the methods presented
- Assess the quality of your sequencing run
- Process reads and align them to a reference genome
- Analyse and interpret long-read sequencing data for variant and stractural variant calling

# Contents

1. [Introduction to ONT](http://who-blackbird.github.io/contents/introduction_to_ONT)
2. Sample preparation
3. [Quality control](http://who-blackbird.github.io/contents/run_quality_control)
4. [Alignment](http://who-blackbird.github.io/contents/alignment)
5. [Variant calling](http://who-blackbird.github.io/contents/variant-calling)
6. [Haplotype phasing](http://who-blackbird.github.io/contents/ONT_course_phasing_Feb2020)
7. [Methylation](http://who-blackbird.github.io/contents/methylation)

# Registration

<a href="https://www.training.cam.ac.uk/event/3327123">https://www.training.cam.ac.uk/event/3327123</a>
